### ISSUE TITLE
Docs: example GitHub Actions configuration with CI + a fuzzing cronjob with shared database

### ISSUE THREAD

**Issue by Zac-HD** [Zac-HD/hypofuzz#14]
I suspect that a copy-pasteable example configuration would get many people over the line from "vaguely interested" to "actually fuzzing", and so this should be a pretty high priority.  I also expect it's a little fiddly to get database sharing set up and working properly between different jobs, so implementing and testing that seems pretty valuable as a FUD-reducer.

See also [Nelson's workflow request](https://blog.nelhage.com/post/property-testing-like-afl/#the-workflow-i-want), [bug-finding vs regression-check testing](https://blog.nelhage.com/post/two-kinds-of-testing/), discussion in https://github.com/Zac-HD/hypofuzz/issues/13#issuecomment-1315903277.

**Comment by agucova** [comment]
I'm thinking about implementing this as:

- A GitHub Actions workflow for code-change fuzzing, running a small length (i.e. 10 minutes) session on every PR
- A GitHub Actions workflow for batch fuzzing, running for a longer session (i.e. 5 hours) in a given frequency (via Github Actions `on.schedule`, instead of a cronjob running somewhere else)

The obvious issue is runner time limits in GitHub Actions, but you can always just move execution to a CI server by setting it up as a self-hosted runner.

As for the shared database, I initially thought as implementing it as [artifacts](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts) for GitHub Actions. However, it seems like it would still make sense to set up a networked database for developer machines to access the database read-only. 

I'm guessing the simpler deployment for that latter case would involve setting up something like a `MultiplexedExampleDatabase` with the CI using `RedisExampleDatabase`, which is seems straightforward enough to run in most servers.

I'll give this a try and report back. I'll use [shed](https://github.com/Zac-HD/shed) as a reference project.

**Comment by Zac-HD** [comment]
Sounds great!  I'm excited to see how this goes 😁 

re: database, consider writing a custom "database" which just grabs the latest artifact for a repo, and gives you read-only access to it?  Multiplex that with a standard directory-based DB and you'll write back, which I presume would then be uploaded by the Actions infrastructure.

The key idea here is that we'd prefer to have an option that requires _no_ additional infrastructure, even if it's a bit janky, because that's so much easier to adopt.  Fine to _also_ have some really nice options which rely on a database though, of course, and we can ship some extra database helpers in Hypothesis itself if that would help.

**Comment by agucova** [comment]
> re: database, consider writing a custom "database" which just grabs the latest artifact for a repo, and gives you read-only access to it? Multiplex that with a standard directory-based DB and you'll write back, which I presume would then be uploaded by the Actions infrastructure.

Oh, that makes a lot of sense! And it doesn't sound particularly difficult (though I might imagine authentication for private repos to be a particular issue).

The other thing I was thinking about is how to get the actual reports from Hypofuzz. Ideally, the output format for `--no-dashboard` would allow for easy extraction of relevant failures, but this currently doesn't seem to be the case. I could prototype some standardized logging format for this.

**Comment by agucova** [comment]
Oh, I'm guessing we could just aggregate the final `Report` for each node?

**Comment by agucova** [comment]
Which then leaves me thinking about how to actually _finish_ the session. Because hypofuzz doesn't seem to support graceful shutdowns, it doesn't seem like it's a good idea to do this externally (i.e. `timeout 600 hypothesis fuzz`). This could happen internally, receiving a `timeout` CLI parameter or something like that.

**Comment by Zac-HD** [comment]
Reporting is designed to all go through the database, so we store data there and you get the report by replaying failing examples using Hypothesis afterwards. (though see #3 for the future architecture here).  

For the same reason, everything is designed to gracefully tolerate being killed, so `timeout 600 hypofuzz` is *exactly* how I'd do it 😁

**Comment by agucova** [comment]
> Reporting is designed to all go through the database, so we store data there and you get the report by replaying failing examples using Hypothesis afterwards. (though see https://github.com/Zac-HD/hypofuzz/issues/3 for the future architecture here).

Oh, this actually makes complete sense! But do you still think it makes sense to report to the user when a failure has happened (i.e. via an issue)?

> For the same reason, everything is designed to gracefully tolerate being killed, so timeout 600 hypofuzz is exactly how I'd do it! The term to search is "crash-only software" 😁

I think this is beautiful ;)

**Comment by Zac-HD** [comment]
> do you still think it makes sense to report to the user when a failure has happened (i.e. via an issue)?

Hmm, probably not?  Typical workflows for teams running this kind of setup are probably more like "run CI hourly on latest master, post to a chat channel on failure".  One issue per failing test would also get massively spammy if you accidentally broke master and got O(thousands) of issues opened, but if it's per-build then the issue is pretty redundant.

**Comment by agucova** [comment]
Update: The artifact-based database works! You can just:
```python
import os

from hypothesis import settings
from hypothesis.database import (DirectoryBasedExampleDatabase,
                                 MultiplexedDatabase)

from hypofuzz.database import GitHubArtifactDatabase

local = DirectoryBasedExampleDatabase(".hypothesis/examples")
shared = GitHubArtifactDatabase("agucova", "hypofuzz")

settings.register_profile("ci", database=local)
settings.register_profile(
    "dev", database=MultiplexedDatabase(local, shared)
)
settings.load_profile("ci" if os.environ.get("CI") else "dev")
```

Add the env variable `GITHUB_TOKEN`, and it works!

**Comment by agucova** [comment]
Now I just need to reduce the workflow to a minimum and document everything ;)

**Comment by agucova** [comment]
@Zac-HD would you prefer `GitHubArtifactDatabase` living on Hypothesis or on Hypofuzz?

If the first, I'll make a PR to Hypothesis first.

**Comment by Zac-HD** [comment]
Let's add it to Hypothesis!

**Comment by agucova** [comment]
I still need some finished for the PR to be merge-worthy, but the minimal workflow [pretty much works](https://github.com/agucova/hypofuzz/actions/runs/4132960127/jobs/7142283523):

```yaml

- name: Download example database
  uses: dawidd6/action-download-artifact@v2.24.3
  with:
    name: hypofuzz-example-db
    path: .hypothesis/examples

- name: Run hypofuzz session
  continue-on-error: true
  run: |
    timeout --preserve-status 20 hypothesis fuzz --no-dashboard

- name: Upload example database as an artifact
  uses: actions/upload-artifact@v3
  with:
    name: hypofuzz-example-db
    path: .hypothesis/examples

```

I still can't nail down why timeout seems to insist in overlaying its own status codes (even with `--preserve-status`) but it should be easy to fix.

**Comment by agucova** [comment]
@Zac-HD I've tested the `GitHubArtifactDatabase` using Hypofuzz, and I've noticed what appears to be a race condition.

When multiprocessing is enabled, more than one artifact is fetched on database initialization. I guess this is natural given how hypofuzz runs. I could formulate some sort of filesystem-based lock for the database initialization process, but I'm not sure if this is something we should handle on the Hypothesis side.

What do you think?

**Comment by Zac-HD** [comment]
I think we should probably handle this on the _HypoFuzz_ side; it'll be easy enough to have a thread that does `settings.default.database.fetch(b"")` parallel to test collection, and then only fork off the multiple processes afterwards.  

In the case where we're _not_ using an exotic setup like this that's almost free; and if we _are_ then it's an optimization overall.  Happily, it also requires no changes at all to the new database implementation 😁

**Comment by agucova** [comment]
I've finished testing and documenting the example workflow on the Hypothesis side (see https://github.com/HypothesisWorks/hypothesis/pull/3636), so I guess the only pending task is fixing the race condition.

I did notice however that Hypofuzz seems incompatible with newer Hypothesis versions, and I'm guessing that's because of changes to Hypothesis's internals.

**Comment by Zac-HD** [comment]
Yep, internal changes in https://github.com/HypothesisWorks/hypothesis/pull/3621 have broken HypoFuzz 😬 

I'll try to fix that early next week, if you don't get to it before me.

**Comment by Zac-HD** [comment]
Following #25, it'd be great to include uploading `.hypothesis/patches/latest_hypofuzz_*.patch` as another artifact.

**Comment by agucova** [comment]
@Zac-HD Would you want to support this directly at the database level?

**Comment by agucova** [comment]
@Zac-HD Should the default behaviour of the database be to just synchronize all patches?

**Comment by Zac-HD** [comment]
The patches aren't in the database (`.hypothesis/examples/` by default), but rather in `.hypothesis/patches/`.  I think we should only upload those matching the glob pattern above; the hash-based names are in this case unhelpfully non-indicative of what's actually in them (failing, covering, or all examples).
